### PR TITLE
Fix several say/think bugs

### LINKED ISSUE #728: Say/think bubbles stop moving with sprite
### Expected Behavior

Say / Think bubbles always stay attached to the sprite they originated from (See: #691)

### Actual Behavior

Say / Think bubble can become fixed in their position and not move with their sprite / clone

### Steps to Reproduce

* Create a stack: When Green Flag Clicked -> Move (10) steps -> Say "Hello!" for (2) seconds.
* Click the green flag a couple of times
* Move the sprite and watch the say block stay where it was.

image

If it helps, make the delay larger from 2 seconds :). I kind of feel the issue is to do with the promise on a timer and a new say block being instanced which the promise is still active? Perhaps the promise needs cancelling when a new say or think block is executed? However, playing around with this, it is really rather unpredictable :)

Also getting this exception in the console:

image

This test setup shows how the say until needs cancelling after another say:

image

And it also has the same issue with the "say for 2" not moving with the sprite, but the second "say for 2" actually moving ok weirdly...

### Operating System and Browser

Chrome latest on windows 10

### PR DESCRIPTION
**Pull request by paulkaplan**
### Resolves

_What Github issue does this resolve (please include link)?_

Resolves https://github.com/LLK/scratch-vm/issues/728

@griffpatch thanks very much for reporting those issues. This PR contains commits to fix that memory leak with event listeners, fixes up the show/hide timeout through green flag clicks and fixes the moving bug. Tiny fixes individually, but together led to some very strange behavior.

Thank you for the thorough repro instructions, that really helps with these bugs.  

For the reviewer, this is one you want to review one commit at a time. The commit description contains more info for each fix.

### CODE REVIEW

**Review by rschamp — CHANGES_REQUESTED** [review]
Discussed offline, there's still some "unique" visibility behavior, so @paulkaplan is going to look into refactoring the handling of the `visible` state.

**Comment by paulkaplan** [comment]
@rschamp thanks for talking through this. I removed the `visible` state, but then realized what it was originally for: it is an optimization specifically targeted at making "typewriter" style projects work without "flashing" of the bubble by removing it and recreating it. 

I added that back in in the last commit, but what I think is a more stable and sane way. The optimization is only used if the skin has already been initialized, so the relevant code is confined to the single "update" branch of the `renderBubble` function.

**Review by rschamp — APPROVED** [review]

File: src/blocks/scratch3_looks.js [hunk 1]
```diff
@@ -6,9 +6,10 @@ const RenderedTarget = require('../sprites/rendered-target');
  * @typedef {object} BubbleState - the bubble state associated with a particular target.
  * @property {Boolean} onSpriteRight - tracks whether the bubble is right or left of the sprite.
  * @property {?int} drawableId - the ID of the associated bubble Drawable, null if none.
+ * @property {Boolean} drawableVisible - if drawable has been hidden by blank text.
+ *      See _renderBubble for explanation of this optimization.
```
Comment (cwillisf) [inline src/blocks/scratch3_looks.js]: Based on the description here it sounds like `drawableVisible` being true means that the Drawable is hidden. Based on the name of the property and the logic in `_renderBubble` it looks like `drawableVisible` being true means that the Drawable is visible. Unless I'm misunderstanding something, could you adjust for consistency? (Maybe just `false if drawable has been hidden by blank text.`?)

**Review by cwillisf — CHANGES_REQUESTED** [review]
One nitpick / misunderstanding, but overall LGTM!

File: src/blocks/scratch3_looks.js (hunk 1 above)
Comment (paulkaplan) [reply src/blocks/scratch3_looks.js]: Oh good catch, sorry I went back and forth adjusting these property names about a dozen times, must have got messed up in the shuffle. Fixed in c14155c

**Pull request merged**